### DISCUSSION TITLE
How to invalidate affected providers by mutation?

### DISCUSSION THREAD

**Discussion by w0rsti** [rrousselGit/riverpod#3589] [Q&A]
Hey guys, thanks for trying to help me! 💯 

Currently developing an app that has an entity called CarBooking which looks like this:

```dart
/// A booking for a car.
@freezed
class CarBooking with _$CarBooking {
  const CarBooking._();

  /// Default constructor a new car booking.
  const factory CarBooking({
    /// The unique identifier of the car booking.
    String? id,

    /// The [Car] this booking is for.
    required Car car,

    /// The [Order] this booking is for.
    ///
    /// If this is `null`, the booking is not related to an order.
    required Order? order,

    /// The date from which on the car will be blocked for this booking.
    required DateTime bookedFrom,

    /// The date until the car will be blocked for this booking.
    required DateTime bookedUntil,
  }) = _CarBooking;

  /// Create a new car booking from a JSON map.
  factory CarBooking.fromJson(Map<String, dynamic> json) =>
      _$CarBookingFromJson(json);
}
```

Therefore I have a CarBookingProvider (Family) which takes an DateTimeRange as an `DateTimeRange range` argument which fetches the CarBookings that are (partially or fully) contained inside the range:
```dart
class CarBookings extends _$CarBookings {
  CarBookingRepository get _repository =>
      ref.read(carBookingRepositoryProvider);

  @override
  FutureOr<List<CarBooking>> build(DateTimeRange range) async {
    return await _repository.getBookingsInRange(range);
  }

  /// Create a new [CarBooking] and update the list of bookings by
  /// prepending it to the current state.
  Future<CarBooking> createBookingForCar({
    required Car car,
    required DateTimeRange duration,
    Order? order
  }) {
    final booking = CarBooking(
      car: car,
      order: order,
      bookedFrom: range.start,
      bookedUntil: range.end,
    );

    update((current) => [booking, ...current])

    //  ** ----> How can i notify affected providers, not all?**
    ref.invalidate(carBookingsProvider)
  }

  // .. other CRUD Operations ...
}
```

When creating a CarBooking I want to invalidate all CarBookingProviders that might be affected. What would be the riverpod way to do so? 
Calling ref.invalidate on the complete family seems kind of overkill since it would invalidate providers which are not affected by the change and therefore cause unnecessary rebuild/fetch requests.

## **Solutions I considered**
* Creating a private variable in the `car_bookings_provider.dart` file like:
```dart
final _watchedRanges = [];

class CarBookings extends _$CarBookings {
  // ...
}
```
and thus keeping track of currently watched ranges (via adding the range in the build and removing it via the ref.onDispose()).
Then in the `createBooking` function of the notifier I can check that list and decide which ranges should be affected by the new booking and call ref.invalidate(carBookingsProvider(range)) for that.

* Same approach as the private variable, just put it inside a provider to make it testable. Alternatively I could make the `_watchedRanges` variable public and mark it with `@visibleForTesting`

* Creating an "Event" Stream in the repository that can be subscribed to and propagates all the Events (e.g. CRUD Operations). Then I can listen to that stream for each family provider member and the provider itself could decide wether to invalidate itself (or even just add it to its state) 

### Evaluation
The more I think about the problem, the more I tend to create an (private) StreamController that exposes a stream of events and each provider member listens to that. The reasons I tend to that solution: 
* It provides the ability to instead invalidating themselves to update their state
* It is more decoupled
* I can pass around that (singleton) Repository (using e.g. Riverpod itself) and call functions of that repository from anywhere and the subscribed providers can react to that changes.
* Other (unrelated) Providers can listen to CarBooking changes if they are interested or depended on the CarBookings in any way.
* When using a real-time backend (like Firebase, Appwrite, Supabase etc....) I can easily add their events to that stream (making changes from other devices possible)

**Comment by snapsl** [comment]
I don't think riverpod provides a simple solution to your problem out of the box. 

> The more I think about the problem, the more I tend to create an (private) StreamController that exposes a stream of events and each provider member listens to that.

That's a good idea! Overall, try to keep it declarative.

## In general:
Does your app really allow multiple instances of the CarbookingsProviderFamily to exist at the same time? Often the old instances is replaced (disposed) by the new one.

Also
>     update((current) => [booking, ...current])
> 
>     //  ** ----> How can i notify affected providers, not all?**
>     ref.invalidate(carBookingsProvider)

Updating and then invalidating makes the update unnecessary, as I understand it.
For side effects this doc exists. [link](https://riverpod.dev/docs/essentials/side_effects#updating-our-local-cache-to-match-the-api-response)

**Reply by w0rsti** [reply]
> I don't think riverpod provides a simple solution to your problem out of the box.
> 
> > The more I think about the problem, the more I tend to create an (private) StreamController that exposes a stream of events and each provider member listens to that.
> 
> That's a good idea! Overall, try to keep it declarative.
> 
> ## In general:
> Does your app really allow multiple instances of the CarbookingsProviderFamily to exist at the same time? Often the old instances is replaced (disposed) by the new one.
> 
> Also
> 
> > ```
> > update((current) => [booking, ...current])
> > 
> > //  ** ----> How can i notify affected providers, not all?**
> > ref.invalidate(carBookingsProvider)
> > ```
> 
> Updating and then invalidating makes the update unnecessary, as I understand it. For side effects this doc exists. [link](https://riverpod.dev/docs/essentials/side_effects#updating-our-local-cache-to-match-the-api-response)

Thanks a lot for your opinion!

I know that updating and then invalidating the same provider does not make any sense.
It was just a "note", that I want to invalidate other providers that might be affected.
